### PR TITLE
Make parallelism optional

### PR DESCRIPTION
**Pull request by n1t0**
Related to #187 

An attempt at implementing your first idea @epwalsh. I just didn't manage to have `map_with` working with `maybe_par_bridge` but everything else seems to be good!

Let me know what you think! And if you have an idea for the `map_with`, that'd be awesome, I couldn't make sense of the error...

### CODE REVIEW

**Comment by epwalsh** [comment]
Oof, that error message has me stumped.

I do like this solution though! I hope we can get it working.

File: tokenizers/src/utils/parallelism.rs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+pub const ENV_VARIABLE: &str = "TOKENIZERS_PARALLELISM";
+
+/// Check if the TOKENIZERS_PARALLELISM env variable has been explicitly set
+pub fn is_parallelism_configured() -> bool {
+    std::env::var(ENV_VARIABLE).is_ok()
+}
+
+/// Get the currently set value for `TOKENIZERS_PARALLELISM` env variable
+pub fn get_parallelism() -> bool {
+    match std::env::var(ENV_VARIABLE) {
```
Comment (epwalsh) [inline tokenizers/src/utils/parallelism.rs:18]: Is the cost of looking up this environment variable every time significant?

File: bindings/python/src/lib.rs [hunk 2]
```diff
@@ -84,6 +103,15 @@ fn normalizers(_py: Python, m: &PyModule) -> PyResult<()> {
 /// Tokenizers Module
 #[pymodule]
 fn tokenizers(_py: Python, m: &PyModule) -> PyResult<()> {
+    // Register the fork callback
+    #[cfg(target_os = "linux")]
+    unsafe {
+        if !REGISTERED_FORK_CALLBACK {
+            libc::pthread_atfork(None, None, Some(child_after_fork));
```
Comment (epwalsh) [inline bindings/python/src/lib.rs:110]: 💯 💯 🎉 🎉 !!!

Comment (epwalsh) [reply bindings/python/src/lib.rs:110]: TIL

Comment (n1t0) [reply bindings/python/src/lib.rs:110]: I didn't know about it either, and it seems to be working really well. This should let us have a very flawless experience. :tada:

**Comment by n1t0** [comment]
Ok, everything should be good now. `map_with` now also works when using `maybe_par_bridge`.

So, the behavior in Python is now the following:
 - When the library is imported for the first time, we register a listener to make sure we catch a potential `fork`
 - If we detect a `fork` and `TOKENIZERS_PARALLELISM` is not set, we disable the parallelism and warn the user
 - If `TOKENIZERS_PARALLELISM` has been set, we use the provided value

Except for the `fork` part, everything works the same on any OS/platform. The environement variable can be used to disable parallelism in any situation.

I think we might wait for https://github.com/cuviper/rayon-cond/pull/2 to be merged before we merge this, but this is now ready for review!

**Comment by n1t0** [comment]
Nice catch! I think it should work fine now.

**Comment by epwalsh** [comment]
I just confirmed it works! 💯

**Comment by n1t0** [comment]
Ok, let's merge this to include it in the upcoming release. We'll see what to do with the `git` dependency when we'll release the Rust version.

**Pull request merged**